### PR TITLE
viro expo plugin

### LINKED ISSUE #87: Expo support out of the box

### PR DESCRIPTION
**Pull request by brunokiafuka**
This PR fixes #87 and #86 

- [x] Ensure [docs](https://viro-community.readme.io/docs/integrating-with-expo) are 100% updated with any "gotchas" or special configuration for M1
- [x] Ensure [docs](https://viro-community.readme.io/docs/integrating-with-expo) are 100% accurate to the instructions found [here](https://robertjcolley.github.io/logbook/blog/2022/02/13/expo-starter-kit-instructions)
- [x] Ensure [docs 2](https://viro-community.readme.io/docs/assets#adding-asset-types-to-expo) are 100% updated for expo metro config
- [x] Verify iOS
- [x] Verify iOS via M1 build
- [x] Verify Android

### CODE REVIEW

**Review by robertjcolley — COMMENTED** [review]
Looks great! @brunokiafuka is there any changes to our build script we need to make to make sure any plugin changes are built and placed in dist? or should the `build` command take care of this?

**Comment by robertjcolley** [comment]
Screen Shot 2022-02-01 at 7 24 53 AM

```
› Planning build
› Executing expo-constants Pods/EXConstants » [CP-User] Generate app.config for prebuilt Constants.manifest
› Executing expostarterkit » Start Packager on http://localhost:8081
› Linking   expostarterkit » expostarterkit
❌  Undefined symbols for architecture x86_64
┌─ Symbol: VROARDeclarativeSession::addARImageTarget(std::__1::shared_ptr<VROARImageTarget>)
└─ Referenced from: ___41-[VRTARImageMarker getARTargetShouldAdd:]_block_invoke_2 in libViroReact.a(VRTARImageMarker.o)


❌  ld: symbol(s) not found for architecture x86_64



❌  clang: error: linker command failed with exit code 1 (use -v to see invocation)



› 3 error(s), and 0 warning(s)

Failed to build iOS project. "xcodebuild" exited with error code 65.
```

I've checked out the PR, and done the following steps to no success. @brunokiafuka am I doing this correctly?

1. expo init expo-starter-kit
2. yarn add @viro-community/react-viro
3. added @viro-community/react-viro to app.json plugins array
4. ran `expo run:ios`
5. error above

Its entirely possible that enabling bitcode in #78 caused this to change your behavior - Charles mentioned that you enable it.

**Comment by robertjcolley** [comment]
https://github.com/robertjcolley/expo-viro-starter-kit is what I've done

File: plugins/withViroIos.ts [hunk 1]
```diff
@@ -16,9 +18,7 @@ const withViroPods = (config: ExpoConfig) => {
 
       fs.readFile(`${root}/Podfile`, "utf-8", (err, data) => {
         data = insertLinesHelper(
-          `  pod 'ViroReact', :path => '../node_modules/@viro-community/react-viro/ios/'
-    pod 'ViroKit_static_lib', :path => '../node_modules/@viro-community/react-viro/ios/dist/ViroRenderer/static_lib'
-            `,
+          `pod 'ViroKit', :path => '../node_modules/@viro-community/react-viro/ios/dist/ViroRenderer/'`,
```
Comment (brunokiafuka) [inline plugins/withViroIos.ts:21]: Replace previous `pod` configuration as per this comment 👉🏾 https://github.com/ViroCommunity/viro/issues/56#issuecomment-986437376

**Comment by brunokiafuka** [comment]
> Looks great! @brunokiafuka is there any changes to our build script we need to make to make sure any plugin changes are built and placed in dist? or should the `build` command take care of this?

the current build should command take care of it.

**Comment by robertjcolley** [comment]
I'm still getting this error:

```
› Linking   expostarterkit » expostarterkit
❌  Undefined symbols for architecture x86_64
┌─ Symbol: VROARDeclarativeSession::addARImageTarget(std::__1::shared_ptr<VROARImageTarget>)
└─ Referenced from: ___41-[VRTARImageMarker getARTargetShouldAdd:]_block_invoke_2 in libViroReact.a(VRTARImageMarker.o)


❌  ld: symbol(s) not found for architecture x86_64



❌  clang: error: linker command failed with exit code 1 (use -v to see invocation)



› 3 error(s), and 0 warning(s)

Failed to build iOS project. "xcodebuild" exited with error code 65.
```

This is after the `expo run:ios` command. After this, I wondered if this was something that would never work because it is trying to build viro for the simulator, so I then did:

0. `expo run:ios`
1. `cd ios` (this is the generated ios folder from expo run:ios)
2. `pod install`
3. `xed .`
4. build with xcode, my phone as the target

More descriptive errors from xcode:

<details><summary>More Errors</summary>
<p>
Undefined symbol: VROARDeclarativeSession::addARImageTarget(std::__1::shared_ptr<VROARImageTarget>)
Undefined symbol: VROMatrix4f::rotateY(float)
Undefined symbol: VROMatrix4f::rotateZ(float)
Undefined symbol: VROHDRLoader::loadRadianceHDRTexture(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)
Undefined symbol: VROAnimationGroup::parse(float, float, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >&, std::__1::vector<std::__1::shared_ptr<VROLazyMaterial>, std::__1::allocator<std::__1::shared_ptr<VROLazyMaterial> > >)
Undefined symbol: VROARNode::setPauseUpdates(bool)
Undefined symbol: VROText::~VROText()
Undefined symbol: VROVector4f::set(float, float, float, float)
Undefined symbol: VROText::VROText(std::__1::basic_string<wchar_t, std::__1::char_traits<wchar_t>, std::__1::allocator<wchar_t> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, VROFontStyle, VROFontWeight, VROVector4f, float, VROTextOuterStroke, int, VROVector4f, float, float, VROTextHorizontalAlignment, VROTextVerticalAlignment, VROLineBreakMode, VROTextClipMode, int, std::__1::shared_ptr<VRODriver>)
Undefined symbol: VROText::update()
Undefined symbol: VROPortal::setBackgroundRotation(VROQuaternion)
Undefined symbol: VRONodeCamera::~VRONodeCamera()
Undefined symbol: VRONodeCamera::setFieldOfViewY(float)
Undefined symbol: VRONodeCamera::VRONodeCamera()
Undefined symbol: VROPortalFrame::~VROPortalFrame()
Undefined symbol: VROPortalFrame::VROPortalFrame()
Undefined symbol: VROParticleEmitter::setBlendMode(VROBlendMode)
Undefined symbol: getBlankTexture(VROTextureType)
Undefined symbol: VROScene::getControllerPresenter()
Undefined symbol: VROPhysicsWorld::findCollisionsWithShape(VROVector3f, VROVector3f, std::__1::shared_ptr<VROPhysicsShape>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)
Undefined symbol: VRONode::getAnimation(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool)
Undefined symbol: VROARDeclarativeSession::addARObjectTarget(std::__1::shared_ptr<VROARObjectTarget>)
Undefined symbol: VROARDeclarativeSession::removeARNode(std::__1::shared_ptr<VROARDeclarativeNode>)
Undefined symbol: VRONode::setScalePivot(VROMatrix4f)
Undefined symbol: VROARScene::setPointCloudSurfaceScale(VROVector3f)
Undefined symbol: VROARScene::displayPointCloud(bool)
Undefined symbol: vtable for VROAnimationChain
Undefined symbol: VROARDeclarativeSession::addARNode(std::__1::shared_ptr<VROARDeclarativeNode>)
Undefined symbol: typeinfo for VROScene
Undefined symbol: VROMaterial::updateSubstrate()
Undefined symbol: typeinfo for VROARScene
Undefined symbol: VROARScene::initDeclarativeSession()
Undefined symbol: VROARScene::setDelegate(std::__1::shared_ptr<VROARSceneDelegate>)
Undefined symbol: VROAnimatedTextureOpenGL::VROAnimatedTextureOpenGL(VROStereoMode)
Undefined symbol: VROARScene::setAnchorDetectionTypes(std::__1::set<VROAnchorDetection, std::__1::less<VROAnchorDetection>, std::__1::allocator<VROAnchorDetection> >)
Undefined symbol: VROARDeclarativeSession::removeARImageTarget(std::__1::shared_ptr<VROARImageTarget>)
Undefined symbol: VROARDeclarativeSession::setDelegate(std::__1::shared_ptr<VROARDeclarativeSessionDelegate>)
Undefined symbol: VROAnimatedTextureOpenGL::loadAnimatedSourceAsync(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<VRODriver>, std::__1::shared_ptr<VROFrameSynchronizer>, std::__1::function<void (bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)>)
Undefined symbol: VROARShadow::apply(std::__1::shared_ptr<VROMaterial>)
Undefined symbol: VROPhysicsBody::setInertia(VROVector3f)
Undefined symbol: VROAnimatedTextureOpenGL::pause()
Undefined symbol: VROScene::~VROScene()
Undefined symbol: vtable for VROSceneController
Undefined symbol: VROPhysicsBody::setIsSimulated(bool)
Undefined symbol: VROScene::VROScene()
Undefined symbol: VROVideoTextureiOS::VROVideoTextureiOS(VROStereoMode, bool)
Undefined symbol: VROARScene::resetPointCloudSurface()
Undefined symbol: VROPortal::setBackgroundCube(std::__1::shared_ptr<VROTexture>)
Undefined symbol: VRONode::setScene(std::__1::shared_ptr<VROScene>, bool)
Undefined symbol: VROPortal::setBackgroundCube(VROVector4f)
Undefined symbol: VROVector3f::normalize() const
Undefined symbol: VRONode::~VRONode()
Undefined symbol: VROPhysicsWorld::setGravity(VROVector3f)
Undefined symbol: VROScene::getRootNode()
Undefined symbol: VROFBXLoader::loadFBXFromResource(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, VROResourceType, std::__1::shared_ptr<VRONode>, std::__1::shared_ptr<VRODriver>, std::__1::function<void (std::__1::shared_ptr<VRONode>, bool)>)
Undefined symbol: VROARObjectTargetiOS::~VROARObjectTargetiOS()
Undefined symbol: VROMorpher::setWeightForTarget(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, float, bool)
Undefined symbol: VRONode::getAnimationKeys(bool)
Undefined symbol: VROSphere::createSphere(float, int, int, bool)
Undefined symbol: VROGLTFLoader::loadGLTFFromResource(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >, VROResourceType, std::__1::shared_ptr<VRONode>, bool, std::__1::shared_ptr<VRODriver>, std::__1::function<void (std::__1::shared_ptr<VRONode>, bool)>)
Undefined symbol: VRONodeCamera::setOrbitFocalPoint(VROVector3f)
Undefined symbol: VRONode::setRotationPivot(VROMatrix4f)
Undefined symbol: _OBJC_CLASS_$_VROViewAR
Undefined symbol: VRONode::setScale(VROVector3f)
Undefined symbol: VROSurface::setHeight(float)
Undefined symbol: VRONode::getLastWorldTransform() const
Undefined symbol: VRONode::deleteGL()
Undefined symbol: VROARDeclarativeSession::removeARObjectTarget(std::__1::shared_ptr<VROARObjectTarget>)
Undefined symbol: VROMorpher::getMorphTargetKeys()
Undefined symbol: VROPhysicsWorld::setDebugDrawVisible(bool)
Undefined symbol: VROLight::setAttenuationStartDistance(float)
Undefined symbol: VROVideoTextureiOS::~VROVideoTextureiOS()
Undefined symbol: typeinfo for VROTexture
Undefined symbol: VROLight::VROLight(VROLightType)
Undefined symbol: VROGeometry::updateSubstrate()
Undefined symbol: VROPhysicsWorld::VROPhysicsWorld()
Undefined symbol: VROMaterialVisual::setTexture(std::__1::shared_ptr<VROTexture>)
Undefined symbol: typeinfo for VROVideoTextureiOS
Undefined symbol: VROMaterial::setChromaKeyFilteringColor(VROVector3f)
Undefined symbol: VROPhysicsBody::clearForces()
Undefined symbol: VROMaterial::setChromaKeyFilteringEnabled(bool)
Undefined symbol: VROMaterial::setShininess(float)
Undefined symbol: VROMaterial::setFresnelExponent(float)
Undefined symbol: VROStringUtil::strcmpinsensitive(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
Undefined symbol: VRONode::getUmbrellaBoundingBox() const
Undefined symbol: VROMaterialVisual::setIntensity(float)
Undefined symbol: VROTexture::VROTexture(bool, std::__1::vector<std::__1::shared_ptr<VROImage>, std::__1::allocator<std::__1::shared_ptr<VROImage> > >&, VROStereoMode)
Undefined symbol: VROMatrix4f::extractScale() const
Undefined symbol: VROMatrix4f::extractRotation(VROVector3f) const
Undefined symbol: VROQuaternion::toEuler() const
Undefined symbol: VROMaterial::setTransparency(float)
Undefined symbol: VRONode::onAnimationFinished()
Undefined symbol: vtable for VROBillboardConstraint
Undefined symbol: VROPhysicsBody::setPhysicsShape(std::__1::shared_ptr<VROPhysicsShape>)
Undefined symbol: VROData::~VROData()
Undefined symbol: vtable for VROGeometry
Undefined symbol: VROBoundingBox::VROBoundingBox()
Undefined symbol: VROData::VROData(void*, int, VRODataOwnership)
Undefined symbol: VROPhysicsShape::~VROPhysicsShape()
</p>
</details>

I see that its for x86_64 arch - I'm wondering if that has anything to do with it. I'll keep poking around with the build config and see if I can get it to work on my phone. Also is there another way to generate this ios folder with expo instead of doing `expo run:ios`?

**Comment by doranteseduardo** [comment]
> I'm still getting this error:
> 
> ```
> › Linking   expostarterkit » expostarterkit
> ❌  Undefined symbols for architecture x86_64
> ┌─ Symbol: VROARDeclarativeSession::addARImageTarget(std::__1::shared_ptr<VROARImageTarget>)
> └─ Referenced from: ___41-[VRTARImageMarker getARTargetShouldAdd:]_block_invoke_2 in libViroReact.a(VRTARImageMarker.o)
> 
> 
> ❌  ld: symbol(s) not found for architecture x86_64
> 
> 
> 
> ❌  clang: error: linker command failed with exit code 1 (use -v to see invocation)
> 
> 
> 
> › 3 error(s), and 0 warning(s)
> 
> Failed to build iOS project. "xcodebuild" exited with error code 65.
> ```
> 
> This is after the `expo run:ios` command. After this, I wondered if this was something that would never work because it is trying to build viro for the simulator, so I then did:
> 
> 0. `expo run:ios`
> 1. `cd ios` (this is the generated ios folder from expo run:ios)
> 2. `pod install`
> 3. `xed .`
> 4. build with xcode, my phone as the target
> 
> More descriptive errors from xcode:
> 
> More Errors
> I see that its for x86_64 arch - I'm wondering if that has anything to do with it. I'll keep poking around with the build config and see if I can get it to work on my phone. Also is there another way to generate this ios folder with expo instead of doing `expo run:ios`?

-x86_64 is simulator on Intel machines

**Comment by robertjcolley** [comment]
Which is what I assumed. Still got the issue when my iPhone 11 was the build target, which I think is arm-based.

**Comment by brunokiafuka** [comment]
@robertjcolley you can try: 
- `expo prebuild --clean -p ios --no-install` to generate the native files, 
- then run `expo run:ios`.

I'll try cloning your starter kit then re-test the plugins using it. Did you run `yarn add ../viro` to install the latest plugin version?

**Comment by robertjcolley** [comment]
Yes I did

**Comment by robertjcolley** [comment]
@brunokiafuka I was able to get the app to build to my phone using the expo project! I had to update the metro.config.js like this to resolve extra assets included (.obj files, etc.) in my project.

```
// metro.config.js
// Learn more https://docs.expo.io/guides/customizing-metro
const { getDefaultConfig } = require("expo/metro-config");

module.exports = (async () => {
  const {
    resolver: { assetExts },
  } = await getDefaultConfig(__dirname);

  return {
    transformer: {
      getTransformOptions: async () => ({
        transform: {
          experimentalImportSupport: false,
          inlineRequires: true,
        },
      }),
    },
    resolver: {
      assetExts: [
        ...assetExts,
        "obj",
        "mtl",
        "mp3",
        "JPG",
        "vrx",
        "hdr",
        "gltf",
        "glb",
        "bin",
        "arobject",
        "gif",
      ],
    },
  };
})();

```

1. `expo prebuild --clean -p ios --no-install`
2. `cd ios && xed . && pod install`
3. build with xcode with phone as target
4. App builds!

However, a new issue is that the app is now crashing when I mount any viro components.

**Comment by robertjcolley** [comment]
@brunokiafuka @doranteseduardo how do we want to handle item 6 of the plugin for expo?

https://github.com/ViroCommunity/viro/blob/main/readmes/INSTALL_ANDROID.md

right now, it looks like AR is the only thing supported with a minimal set of permssions/features. Should we enable VR by default as well? Should we have a guide to adding a plugin which enables VR in expo on android?

**Comment by brunokiafuka** [comment]
> That part is made by the plug-in itself.
> 
> Edit:
> A package for each XR option seems the most feasible but I'm unsure if it's the best approach.

We can pass the option as a plug-in prop abs then edit it accordingly.

**Comment by robertjcolley** [comment]
> > That part is made by the plug-in itself.
> > 
> > Edit:
> > A package for each XR option seems the most feasible but I'm unsure if it's the best approach.
> 
> We can pass the option as a plug-in prop abs then edit it accordingly. 

I think that makes the most sense - maybe an array in app.json or app.config.js.

File: plugins/withViroAndroid.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+          (plugin) =>
+            Array.isArray(plugin) && plugin[0] === "@viro-community/react-viro"
+        );
+        let viroPluginConfig = "AR";
+        if (
+          Array.isArray(viroPlugin) &&
+          ["AR", "GVR", "OVR_MOBILE"].includes(viroPlugin[1]?.androidXrMode)
+        ) {
+          viroPluginConfig = viroPlugin[1]?.androidXrMode;
+        }
```
Comment (robertjcolley) [inline plugins/withViroAndroid.ts:77]: @doranteseduardo @brunokiafuka I've added some code which allows for the GVR/AR/OVR_MOBILE option in app.json.

**Comment by robertjcolley** [comment]
I'm considering this PR close to mergeable! See thread here. Waiting on testing!

https://discord.com/channels/774471080713781259/884270800180502538/942505347728437288

File: plugins/withViroAndroid.ts (hunk 2 above)
Comment (robertjcolley) [reply plugins/withViroAndroid.ts:77]:
```javascript
        const viroPlugin = config?.plugins?.find(
          (plugin) =>
            Array.isArray(plugin) && plugin[0] === "@viro-community/react-viro"
        );
```

should we throw an error if the plugin is not in app.json? do we need to worry about that?

File: dist/plugins/withViroAndroid.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                 *     "@viro-community/react-viro",
+                 *     {
+                 *       "androidXrMode": "GVR"
+                 *     }
+                 *   ]
+                 * ],
+                 *
+                 * ********************************************************************
+                 * Sample app.json with multiple options for Viro config
+                 * The default configuration is "AR"
```
Comment (robertjcolley) [inline dist/plugins/withViroAndroid.js:39]:
```suggestion
```

File: plugins/withViroAndroid.ts [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+         *   [
+         *     "@viro-community/react-viro",
+         *     {
+         *       "androidXrMode": ["GVR", "AR"]
+         *     }
+         *   ]
+         * ],
+         * ********************************************************************
+         * Sample app.json without property config
+         * The default configuration is "AR"
```
Comment (robertjcolley) [inline plugins/withViroAndroid.ts:72]:
```suggestion
```

**Review by robertjcolley — DISMISSED** [review]

**Review by robertjcolley — APPROVED** [review]

**Comment by haibert** [comment]
This is awesome will this be added to the [expo config plugin repo](https://github.com/expo/config-plugins)?

**Comment by robertjcolley** [comment]
> This is awesome will this be added to the [expo config plugin repo](https://github.com/expo/config-plugins)?

I believe it is!

**Pull request merged**